server/core/src/irods_resource_manager.cpp

```diff
@@ -13,226 +14,213 @@
     } // validate_vault_path
 
-// =-=-=-=-=-=-=-
-// public - connect to the catalog and query for all the
-//          attached resources and instantiate them
-    error resource_manager::init_from_catalog(rsComm_t& _comm)
+    void resource_manager::init_resource_from_database_row(const std::vector<std::string>& _row)
+    {
+        const auto& name = _row.at(1);
+        const auto& type = _row.at(3);
+        const auto& context = _row.at(14);
+
+        resource_ptr resc_ptr;
+        if (const auto ret = load_resource_plugin(resc_ptr, type, name, context); !ret.ok()) {
+            irods::log(PASS(ret));
+            return;
+        }
+
+        const auto& location = _row.at(5);
+        if (irods::EMPTY_RESC_HOST != location) {
+            const auto& zone_name = _row.at(2);
+            rodsHostAddr_t addr{};
+            std::strncpy(addr.hostAddr, location.c_str(), LONG_NAME_LEN);
+            std::strncpy(addr.zoneName, zone_name.c_str(), NAME_LEN);
+
+            rodsServerHost_t* host{};
+            if (resolveHost(&addr, &host) < 0) {
+                log_resc::info("[{}:{}] - resolveHost error for [{}]", __func__, __LINE__, addr.hostAddr);
+            }
+
+            resc_ptr->set_property<rodsServerHost_t*>(RESOURCE_HOST, host);
+        }
+        else {
+            resc_ptr->set_property<rodsServerHost_t*>(RESOURCE_HOST, nullptr);
+        }
+
+        const rodsLong_t id = std::strtoll(_row.at(0).c_str(), 0, 0);
+        resc_ptr->set_property<rodsLong_t>(RESOURCE_ID, id);
+        resc_ptr->set_property<long>(RESOURCE_QUOTA, RESC_QUOTA_UNINIT);
+
+        const auto& status = _row.at(12);
+        resc_ptr->set_property(RESOURCE_STATUS, (status == RESC_DOWN) ? INT_RESC_STATUS_DOWN : INT_RESC_STATUS_UP);
+
+        resc_ptr->set_property<std::string>(RESOURCE_FREESPACE,      _row.at(7));
+        resc_ptr->set_property<std::string>(RESOURCE_ZONE,           _row.at(2));
+        resc_ptr->set_property<std::string>(RESOURCE_NAME,           name);
+        resc_ptr->set_property<std::string>(RESOURCE_LOCATION,       location);
+        resc_ptr->set_property<std::string>(RESOURCE_TYPE,           type);
+        resc_ptr->set_property<std::string>(RESOURCE_CLASS,          _row.at(4));
+        resc_ptr->set_property<std::string>(RESOURCE_PATH,           _row.at(6));
+        resc_ptr->set_property<std::string>(RESOURCE_INFO,           _row.at(8));
+        resc_ptr->set_property<std::string>(RESOURCE_COMMENTS,       _row.at(9));
+        resc_ptr->set_property<std::string>(RESOURCE_CREATE_TS,      _row.at(10));
+        resc_ptr->set_property<std::string>(RESOURCE_MODIFY_TS,      _row.at(11));
+        resc_ptr->set_property<std::string>(RESOURCE_CHILDREN,       _row.at(13));
+        resc_ptr->set_property<std::string>(RESOURCE_CONTEXT,        context);
+        resc_ptr->set_property<std::string>(RESOURCE_PARENT,         _row.at(15));
```

Comment:
We need to find a better way to deal with the resource id and parent. They are both integer values, but one is stored as an integer while the other as a string.

Perhaps we should move away from map-like representations for certain things.

The following is easier to handle as compared to a map-like type.
```cpp
struct resource
{
    int id;
    optional<int> parent_id;
    string name;
    string location;
    string vault_path;
};
```
This makes even more sense given that this information is strictly for server-side use.

Reply:
agreed... but not quite sure what you mean with server-side use.

we definitely send this information to things like ilsresc and izonereport

Reply:
The interface used to set/get info about a specific resource requires that you know the type of the object and requires verbose function calls to access it. For example:
```cpp
// (1) Lookup the resource pointer by resource name.
auto resc_ptr = resource_name_map[resc_name_string];

// (2) Extract the resource id.
rodsLong_t resc_id;
auto error = resc_ptr->get<rodsLong_t>(irods::RESOURCE_ID, resc_id);
if (!error.ok()) {
    // Something bad happened. Can't move forward.
}

// (3) We have the resource id. Do work!
```
The step of interest is (2). It takes several lines to fetch this information and can lead to time spent in the debugger if the type passed to `get<T>` is incorrect. This design allows arbitrary objects to be attached to the resource, but leads to bugs.

Given that there is a set of properties known to be attached to every resource, we can replace this generic design with a simpler type. Hence the example in my previous comment. The types are concrete and the dev doesn't have to risk getting stuck trying to figure out the type of the stored property.

This can be achieved without having to replace the original design. We can introduce a set of free functions that do the work for use. For example:
```cpp
auto resource_id(const resource& _r) -> rodsLong_t;
auto resource_parent_id(const resource& _r) -> std::optional<std::string>;
```
All of the work is now hidden behind a clean interface. I think this is the quickest and simplest solution for cleaning up resource manager/pointer code.

I'll go ahead and implement this and adjust the resource manager implementation to make use of it.

By server-side use, I mean the iRODS data types that store this info are only available and used in server-side code. We need to make accessing that info less error prone and verbose.